docs/how-to/incremental-migrations.md

```diff
@@ -0,0 +1,365 @@
+
+$ spacetime call -s local incr-migration-demo level_up_character
+
+2025-01-07T15:34:01.437495Z  INFO: src/lib.rs:66: Leveling up Gefjon from 1 to 2
+
+$ spacetime sql incr-migration-demo 'SELECT * FROM character'
+
+ player_id | nickname | level | class
+-----------+----------+-------+----------------
+ <snip>    | "Gefjon" | 2     | (Fighter = ()) 
+```
+
+Now we want to add a new feature: each player should be able to align themselves with the forces of good or evil, so we can get some healthy competition going between our players. We'll start each character off with `Alliance::Neutral`, and then offer them a reducer `choose_alliance` to set it to either `Alliance::Good` or `Alliance::Evil`. Our first attempt will be to add a new column to the type `Character`:
+
+```rust
+#[spacetimedb::table(name = character, public)]
+struct Character {
+    #[primary_key]
+    player_id: Identity,
+    nickname: String,
+    level: u32,
+    class: Class,
+    alliance: Alliance,
+}
+
+#[derive(SpacetimeType, Debug, Copy, Clone)]
+enum Alliance {
+    Good,
+    Neutral,
+    Evil,
+}
+
+#[spacetimedb::reducer]
+fn choose_alliance(ctx: &ReducerContext, alliance: Alliance) {
+    let character = find_character_for_player(ctx);
+    log::info!(
+        "Setting {}'s alliance to {:?} for player {}",
+        character.nickname,
+        alliance,
+        ctx.sender,
+    );
+    update_character(
+        ctx,
+        Character {
+            alliance,
+            ..character
+        },
+    );
+}
+```
+
+But that will fail, since SpacetimeDB doesn't know how to update our existing `character` rows with the new column:
+
+```
+Error: Database update rejected: Errors occurred:
+Adding a column alliance to table character requires a manual migration
+```
+
+Instead, we'll add a new table, `character_v2`, which will coexist with our original `character` table:
+
+```rust
+#[spacetimedb::table(name = character_v2, public)]
+struct CharacterV2 {
+    #[primary_key]
+    player_id: Identity,
+    nickname: String,
+    level: u32,
+    class: Class,
+    alliance: Alliance,
+}
+```
+
+When a new player creates a character, we'll make rows in both tables for them. This way, any old clients that are still subscribing to the original `character` table will continue to work, though of course they won't know about the character's alliance.
+
+```rust
+#[spacetimedb::reducer]
+fn create_character(ctx: &ReducerContext, class: Class, nickname: String) {
+    log::info!(
+        "Creating new level 1 {class:?} named {nickname} for player {}",
+        ctx.sender,
+    );
+    ctx.db.character().insert(Character {
+        player_id: ctx.sender,
+        nickname: nickname.clone(),
+        level: 1,
+        class,
+    });
+    ctx.db.character_v2().insert(CharacterV2 {
+        player_id: ctx.sender,
+        nickname,
+        level: 1,
+        class,
+        alliance: Alliance::Neutral,
+    });
+}
```

Comment:
If the database developer is using any `auto_inc` fields in `Character`, they will need to choose between:

1. Creating *new* `auto_inc` fields in `CharacterV2`, which will be out of sync with those in `Character`.
2. Or, making the new fields in `CharacterV2` non-auto-inc, with values mirrored from `Character`.

This pattern is correct for choice 1 and incorrect for choice 2.

Reply:
In general I'm unhappy that there's multiple places where `CharacterV2` is initialized from `Character`. In this case you keep them consistent but it's easy to screw up with more data.

I would recommend instead creating a `Character` here and calling `find_character_for_player`.

Reply:
Hm, you're right that auto_inc fields are complex. I think the correct solution is:
- Any fields that are in both tables should be auto_inc only in v1, and mirrored in v2.
- Any fields that are newly added in v2 should be auto_inc there.

> In general I'm unhappy that there's multiple places where CharacterV2 is initialized from Character. In this case you keep them consistent but it's easy to screw up with more data.

What do you mean? There's only one place where a `character` row gets translated to `character_v2` in this code. There's also one place where new rows get created, and one place where rows get updated.

Reply:
https://github.com/orgs/clockworklabs/projects/22/views/16?pane=issue&itemId=92869068 would be even better for this in the future.

Reply:
`CharacterV2`s are initialized at both 199:205 and 225:230. It's a different flow when creating a fresh `CharacterV2` versus initializing a `CharacterV2` backwards-compatibly from a `CharacterV1`.

Really it's a case-by-case issue. It reminds me of lightfox's chains of migrations. When initializing a new row of an incrementally migrated table, do you:
- Create the newest version of the table, and then write-back to the old versions?
- Or, create the oldest version of the table, and then run your migration chain to default-initialize the new data?

If you have auto-inc fields in the old table, you need to start old and then roll your changes forward. If you don't, you can do whatever.

Reply:
The happy medium is:

- Create the oldest version of the row, run your migration chain to initialize new versions of the row, then call the write-back function on the newest version.

That is:
```
fn create_character(ctx: &ReducerCtx) {
    let v1 = character.insert(Character { /* ... */ });
    let v2 = find_character(ctx, v1.id);
    update_character(ctx, v2);
}
```

This means that the same initialization logic happens for all rows of the new table. It also means that there's a single point of enforcement for the sync logic (`update_character`), and this logic has always been run for all rows of the new table.

Reply:
I've removed the `auto_inc` field. I think this document as it stands now is enough to set people in the right direction, at least, and we can add more in the future.